Create notebook introducing data collections

emmarogge opened the pull request:
This notebook contains widgets to assist users in creating a workspace then converting it to a data collection. It also provides instructions and an embedded screencast of how to add the data collection to a Terra workspace.

emmarogge committed:
Create notebook with widgets that facilitate creation of data collection.

wnojopra commented on first_hour_on_terra/README.md:
```diff
@@ -0,0 +1,10 @@
+
+## Getting started
+
+Run the [working_with_data_collections.ipynb](https://github.com/todo) notebook to learn about creating data collections and adding them to your workspaces.
```
Fix the todo link

emmarogge replied:
Yes, once this PR merged in to the main branch of this repo, I will be able to replace that link with the actual URL.

wnojopra replied:
We can figure out that the URL will be https://github.com/DataBiosphere/terra-axon-examples/blob/main/first_hour_on_terra/working_with_data_collections.ipynb . So you can put that in if you wanted to do sooner than later.

emmarogge replied:
updated!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,449 @@
+    "import json\n",
+    "from ipywidgets import Button, Layout\n",
+    "\n",
+    "CURRENT_WORKSPACE_ID = !terra workspace describe --format=json | jq --raw-output \".id\""
```
I tend to prefer notebook cells to output **something**, just to show that it actually did something. Here you could add something like `print(f"Current workspace ID is {CURRENT_WORKSPACE_ID}")`

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,449 @@
+   "source": [
+    "## Overview\n",
+    "\n",
+    "This notebook provides examples of working with data collections in Terra. Build upon the best practices described in this notebook to create and share your own data collections. "
```
I think it'd be worth also mentioning here what a data collection is. Pulling directly from our docs site would be fine.

emmarogge replied:
agreed -- added content & links to documenation!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,449 @@
+    "input_style= {'description_width':'initial'}\n",
+    "buttonOutput = 'Please provide appropriate values in the input boxes.'\n",
+    "\n",
+    "input_workspace_id = widgets.Text(\n",
```
Running this cell I get an import error. I think you're missing `from ipywidgets import widgets`?

emmarogge replied:
fixed!

wnojopra replied:
Instead of a new cell, could you please merge this with the other ipywdigets imports in the first cell?

wnojopra replied:
Can you double check your imports? For example I see you're using `Layout` and `os` without importing them, which causes import errors. What I'd suggest doing is clicking on kernel -> restart kernel and then running the notebook again from the top.

emmarogge replied:
I have refactored this and cleaned up the unnecessary imports.

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,449 @@
+    "1. Click the \"+ Data catalog\" button.\n",
+    "1. Select your newly created data collection from those listed in the modal.\n",
+    "1. Navigate through the steps in the modal to complete the addition of the data collection to your workspace.\n",
+    "1. On a cloud environment terminal in your workspace, run `terra resource list` to confirm the data collection has been added.\n",
```
As an alternative option, we could add a cell here with `!terra resource list` that one could run after adding the data collection.

emmarogge replied:
Since I cannot capture that step in the screencast, I'm a little worried it might be awkward to switch back to Bash commands at this point...what do you think? I am happy to include it but not sure where to put it relative to the screencast.

wnojopra replied:
I'd also be ok if it said "On a cloud environment terminal in your workspace or in a Jupyter cell". Its a minor point, I don't think it blocks this from being merged.

emmarogge replied:
I have added the command as you suggested -- I tested it out and I am no longer concerned it will be awkward after seeing it.

wnojopra replied:
Your instructions said `terra resource list` but your new command says `!terra workspace list`. Which should it be? And could you please also add some markdown briefly mentioning what they should be seeing? Since the cell is divided from the instructions by the video the step "Run the cell below" can be missed.

emmarogge replied:
Per discussion over chat, I've removed this step.

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,449 @@
+    "1. Navigate through the steps in the modal to complete the addition of the data collection to your workspace.\n",
+    "1. On a cloud environment terminal in your workspace, run `terra resource list` to confirm the data collection has been added.\n",
+    "\n",
+    "<video controls src=\"screencasts/add_data_collection_to_workspace.mp4\" width=600>Add data collection to workspace</video>"
```
Cool!

wnojopra reviewed (changes requested):
Looks great! A few small things to fix

emmarogge committed:
Changes based on reviewer feedback

- Created function for CURRENT_WORKSPACE_ID because Terra CLI returns a list.
- Print CURRENT_WORKSPACE_ID to show the cell performed the operation.
- Added more content & links to docs in Overview section.
- Fixed import of ipywidgets package.
- Added cell to check for data collection in resource list & related content.

emmarogge committed:
Refactor in response to reviewer feedback

- Update README link to point to actual notebook in main branch of repo.
- Clean up imports.
- Remove confusing instruction to list workspaces after adding DC to WS in UI.

emmarogge committed:
Use new name (Verily Workbench)

wnojopra commented on first_hour_on_terra/README.md:
```diff
@@ -0,0 +1,10 @@
+
+## Getting started
+
+Run the [working_with_data_collections.ipynb]( https://github.com/DataBiosphere/terra-axon-examples/blob/main/first_hour_on_terra/working_with_data_collections.ipynb) notebook to learn about creating data collections and adding them to your workspaces.
```
Nit: Cleanup the stray space at the start of the URL

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,476 @@
+   "source": [
+    "### Create a new workspace\n",
+    "\n",
+    "In order to create a data collection, you must first create a new workspace. Run the cell below to create a new workspace. The output should resemble:\n",
```
I think it needs to be more clear what the user is expected to input here. Please add instructions here that tells the user that they're expected to create a new workspace id, description, and resource name, and give examples of good names. You have good examples in the expected output, you can reuse those in the instructions.

emmarogge replied:
updated!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,476 @@
+    "  terra-type: data-collection\n",
+    "Created:           2023-05-03\n",
+    "Last updated:      2023-05-03\n",
+    "# Resources:       0Workspace properties successfully updated.\n",
```
Nit: Fix the typo with 0Workspace

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,476 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Convert workspace to data collection\n",
```
Same comment as above about making it clear what the user is expected to input. Please add instructions here that tells the user that they're expected to input the SAME workspace id, a description (is this expected to be same or different?), and what version is expected to be. And again please give examples of good names.

emmarogge replied:
updated!

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,476 @@
+    "        terraCommand = f\"terra workspace create --id={input_workspace_id.value} --description=\\\"{input_description.value}\\\" --name={input_name.value}\"\"\".format(input_workspace_id.value,input_description.value,input_name.value)       \n",
+    "        print(\"Running command to create workspace...\")\n",
+    "        print(terraCommand)\n",
+    "        result = subprocess.run([\"terra\",\"workspace\",\"create\",f\"--id={input_workspace_id.value}\",f\"--description=\\\"{input_description.value}\\\"\",f\"--name={input_name.value}\"],\n",
```
What happens if this command fails? Say, I give a workspace ID that already exists.

emmarogge replied:
I have now updated both widgets to print the success message (from stdout) if the command succeeds, but to otherwise print the error message (from stderr) so that users can learn from the error message what went wrong.

emmarogge committed:
Update based on reviewer feedback

- Fix typo in README
- Provide more information about parameters for workspace creation widget.
- Provide more information about parameters for WS to data collection conversion widget.
- Fix typo in expected output of WS-to-DC conversion widget.
- Update widgets to print stderr if command does not succeed, and to print stdout upon success.

wnojopra commented on first_hour_on_terra/working_with_data_collections.ipynb:
```diff
@@ -227,34 +234,42 @@
+    "Run the cell below to create a widget, then populate the widget's input fields and click the button to convert the workspace to a data collection.\n",
+    "\n",
+    "Widget input parameters include:\n",
+    "- `workspace_id`: Should match the workspace ID of the workspace created above.\n",
```
The workspace_id HAS to match, right? "Must match" instead of Should match.

emmarogge replied:
Fixed!

wnojopra reviewed (approved):
Awesome! One more comment but otherwise LGTM

emmarogge committed:
Workspace ID must match above
